core/src/main/java/net/opentsdb/query/processor/bucketquantile/BucketQuantile.java

```diff
@@ -0,0 +1,387 @@
+            + "] did not match the regex pattern: " + config.getBucketRegex());
+      }
+      if (matcher.groupCount() < 2) {
+        throw new IllegalArgumentException("Histogram metric [" + histogram 
+            + "] did not match two groups from the regex pattern: " 
+            + config.getBucketRegex());
+      }
+      double lower = Double.NaN;
+      try {
+         lower = Double.parseDouble(matcher.group(1));
+      } catch (NumberFormatException e) {
+        // actually shouldn't happen as the regex would toss it
+        throw new IllegalArgumentException("Failed to parse lower bucket from "
+            + "metric [" + histogram + "] using pattern: " 
+            + config.getBucketRegex() + " that matched: " + matcher.group(1));
+      }
+      
+      double upper = Double.NaN;
+      try {
+        // actually shouldn't happen as the regex would toss it
+        upper = Double.parseDouble(matcher.group(2));
+      } catch (NumberFormatException e) {
+        throw new IllegalArgumentException("Failed to parse upper bucket from "
+            + "metric [" + histogram + "] using pattern: " 
+            + config.getBucketRegex() + " that matched: " + matcher.group(2));
+      }
+      buckets[bucket_idx++] = new Bucket(histogram, lower, upper);
+    }
+    
+    if (!Strings.isNullOrEmpty(config.underflowMetric())) {
+      buckets[bucket_idx++] = new Bucket(config.underflowMetric(), false, true);
+    }
+    
+    if (!Strings.isNullOrEmpty(config.overflowMetric())) {
+      buckets[bucket_idx++] = new Bucket(config.overflowMetric(), true, false);
+    }
+    
+    Arrays.sort(buckets, BUCKET_COMPARATOR);
+    ObjectPool pool = context.tsdb().getRegistry().getObjectPool(IntArrayPool.TYPE); 
+    int_array_pool = pool != null ? (ArrayObjectPool) pool : null;
+    pool = context.tsdb().getRegistry().getObjectPool(LongArrayPool.TYPE);
+    long_array_pool = pool != null ? (ArrayObjectPool) pool : null;
+    pool = context.tsdb().getRegistry().getObjectPool(DoubleArrayPool.TYPE);
+    double_array_pool = pool != null ? (ArrayObjectPool) pool : null;
+  }
+
+  @Override
+  public QueryNodeConfig config() {
+    return config;
+  }
+
+  @Override
+  public void close() {
+    // no-op
+  }
+
+  @Override
+  public void onNext(final QueryResult next) {
+    if (results.get(next.dataSource()) != DUMMY) {
+      LOG.warn("Unexpected result: " + next.dataSource());
+      return;
+    }
+    
+    if (!Strings.isNullOrEmpty(next.error()) || next.exception() != null) {
+      // failed
+      if (failed.compareAndSet(false, true)) {
+        if (next.exception() != null) {
+          onError(next.exception());
+        } else {
+          onError(new QueryDownstreamException(next.error()));
+        }
+      }
+    }
+    
+    results.put(next.dataSource(), next);
+    int dummies = 0;
+    for (final Entry<QueryResultId, QueryResult> entry : results.entrySet()) {
```

Comment:
Could CME during for loop if onNext() called from multiple threads.

Reply:
It's a concurrent map so we're safe here as the iterator returns a thread safe view.